vulcan_geometry.py

```diff
@@ -0,0 +1,259 @@
+#!/usr/bin/env python
+from helper import INCH_TO_METRE, MantidGeom
+from lxml import etree as le  # python-lxml on rpm based systems
+import numpy as np
+from rectangle import Rectangle, Vector, makeLocation
+from sns_ncolumn import readFile
+
+L1: float = -43.754  # meter
+TUBE_LENGTH: float = 1.0  # meter
+TUBE_LENGTH_SHORT: float = 0.7  # meters
+TUBE_RADIUS: float = 0.317 * INCH_TO_METRE * 0.5  # given as diameter
+TUBE_PIXELS: int = 512
+PIXELS_PER_EIGHTPACK: int = 8*TUBE_PIXELS  # number of actual pixels
+PIXELS_PER_PANEL: int = 5000  # reserve extra pixels after the end of the 8-pack
+PIXELS_PER_BANK: int = PIXELS_PER_PANEL * 20  # maximum number of 8-pack in a bank
+SEPARATION: float = 0.323 * INCH_TO_METRE  # distance between front and back
+SLIP: float = 0.434 * INCH_TO_METRE  # distance between 2 and 4
+SLIP_PANEL: float = 3 * SLIP + 0.460 * INCH_TO_METRE  # bonus spacing between panel centers
+
+CSV_FILE: str = 'SNS/VULCAN/VULCAN_geom_20210210.csv'
+
+
+def readPositions(filename: str = CSV_FILE):
+    '''The CSV file has measurements of the front tubes of each 8-pack'''
+    # read in and delete unneccessary columns
+    positions = readFile(filename, delimiter=',')
+    del positions['L']  # tube length
+    del positions['C']  # tube centers
+
+    # set up a dict to split the data into
+    banks_allpixels = {'bank1': {},
+                       'bank2': {},
+                       'bank5': {}}
+    for bank_label in banks_allpixels.keys():
+        for column in positions.keys():
+            banks_allpixels[bank_label][column] = []
+
+    # split the data into banks
+    for i, point_label in enumerate(positions['Point']):
+        bank_label = ''
+        if point_label.startswith('HA_'):
+            bank_label = 'bank5'
+        elif point_label.startswith('BR_'):
+            bank_label = 'bank1'
+        elif point_label.startswith('BL_'):
+            bank_label = 'bank2'
+        if bank_label:
+            banks_allpixels[bank_label]['Point'].append(point_label.split('_')[-1])
+            banks_allpixels[bank_label]['X'].append(positions['X'][i])
+            banks_allpixels[bank_label]['Y'].append(positions['Y'][i])
+            banks_allpixels[bank_label]['Z'].append(positions['Z'][i])
```

Comment:
Just a comment here, all of these operations can be simplified if we switch to pandas.dataframe.

Reply:
You are completely correct. I was using library functions that already exist rather than adding a dependency. The thing that scares me with refactoring the code in this repository is how limited and brittle the tests already are.